pkg/chunk/cache/redis_client.go

```diff
@@ -0,0 +1,304 @@
+package cache
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"flag"
+	"fmt"
+	"strings"
+	"time"
+	"unsafe"
+
+	"github.com/cortexproject/cortex/pkg/util/flagext"
+
+	"github.com/go-redis/redis/v8"
+)
+
+const (
+	redisTopologyServer   string = "server"
+	redisTopologyCluster  string = "cluster"
+	redisTopologySentinel string = "sentinel"
+)
+
+var ErrNoMasters = errors.New("redis: no masters")
+
+// RedisConfig defines how a RedisCache should be constructed.
+type RedisConfig struct {
+	Topology    string         `yaml:"topology"`
+	Endpoint    string         `yaml:"endpoint"`
+	Timeout     time.Duration  `yaml:"timeout"`
+	Expiration  time.Duration  `yaml:"expiration"`
+	PoolSize    int            `yaml:"pool_size"`
+	Password    flagext.Secret `yaml:"password"`
+	EnableTLS   bool           `yaml:"enable_tls"`
+	IdleTimeout time.Duration  `yaml:"idle_timeout"`
+	MaxConnAge  time.Duration  `yaml:"max_connection_age"`
+
+	DeprecatedMaxIdleConns         int           `yaml:"max_idle_conns"`
+	DeprecatedMaxActiveConns       int           `yaml:"max_active_conns"`
+	DeprecatedMaxConnLifetime      time.Duration `yaml:"max_conn_lifetime"`
+	DeprecatedWaitOnPoolExhaustion bool          `yaml:"wait_on_pool_exhaustion"`
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *RedisConfig) RegisterFlagsWithPrefix(prefix, description string, f *flag.FlagSet) {
+	f.StringVar(&cfg.Topology, prefix+"redis.topology", redisTopologyServer, description+"Redis topology. Supported: "+redisTopologyServer+", "+redisTopologyCluster+", "+redisTopologySentinel+".")
+	f.StringVar(&cfg.Endpoint, prefix+"redis.endpoint", "", description+"Redis service endpoint to use when caching chunks. If empty, no redis will be used.")
```

Comment:
Please specify that it can be a comma-separated list of endpoints when the topology is cluster.